Readme.md
# http_to_nsq

 Publishes HTTP requests to NSQD.

## Installation

 Binary [releases](https://github.com/segmentio/http_to_nsq/releases), or via:

```
$ go get github.com/segmentio/http_to_nsq/cmd/http_to_nsq
```

## About

 Useful for things like creating an end-point for CI web hooks.

 Currently only JSON bodies are supported, this may be converted to a blob later, but base64 is meh, so for now JSON is it!

## Usage

#### type Message

```go
type Message struct {
	URL    string          `json:"url"`
	Method string          `json:"method"`
	Header http.Header     `json:"header"`
	Body   json.RawMessage `json:"body"`
}
```

# License

MIT


History.md

0.2.0 / 2017-06-05
===================

  * Fix: Respect Topic argument supplied to binary.
  * Improvement: Faster JSON Parsing.


v0.1.0 / 2015-05-12
===================

  * add missing [error] to a log
  * add log for publishing
  * add log levels
  * remove log prefix

v0.0.3 / 2015-05-07
===================

  * add --secret
  * add missing comment
  * add History.md

v0.0.2 / 2015-05-07
===================

  * add /internal/health


server_test.go
package http_to_nsq

import (
	"bytes"
	"io/ioutil"
	"log"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/bmizerany/assert"
)

type pub struct {
	msgs [][]byte
}

func (p *pub) Publish(topic string, body []byte) error {
	p.msgs = append(p.msgs, body)
	return nil
}

func TestServer_ServeHTTP_POST(t *testing.T) {
	p := new(pub)

	s := Server{
		Log:       log.New(ioutil.Discard, "", log.LstdFlags),
		Topic:     "builds",
		Publisher: p,
	}

	b := bytes.NewBufferString(`{ "foo": "bar" }`)

	r, err := http.NewRequest("POST", "/build", b)
	assert.Equal(t, nil, err)
	r.Header.Set("Content-Type", "application/json")

	w := httptest.NewRecorder()
	s.ServeHTTP(w, r)

	assert.Equal(t, 1, len(p.msgs))
	assert.Equal(t, `{"url":"/build","method":"POST","header":{"Content-Type":["application/json"]},"body":{"foo":"bar"}}`, string(p.msgs[0]))

	assert.Equal(t, 200, w.Code)
	assert.Equal(t, ":)", w.Body.String())
}

func TestServer_ServeHTTP_GET(t *testing.T) {
	p := new(pub)

	s := Server{
		Log:       log.New(ioutil.Discard, "", log.LstdFlags),
		Topic:     "builds",
		Publisher: p,
	}

	r, err := http.NewRequest("GET", "/build", nil)
	assert.Equal(t, nil, err)

	w := httptest.NewRecorder()
	s.ServeHTTP(w, r)

	assert.Equal(t, 1, len(p.msgs))
	assert.Equal(t, `{"url":"/build","method":"GET","header":{},"body":null}`, string(p.msgs[0]))

	assert.Equal(t, 200, w.Code)
	assert.Equal(t, ":)", w.Body.String())
}

func TestServer_ServeHTTP_secret_invalid(t *testing.T) {
	p := new(pub)

	s := Server{
		Log:       log.New(ioutil.Discard, "", log.LstdFlags),
		Topic:     "builds",
		Secret:    "wahoo",
		Publisher: p,
	}

	b := bytes.NewBufferString(`{ "foo": "bar" }`)

	r, err := http.NewRequest("POST", "/build", b)
	assert.Equal(t, nil, err)
	r.Header.Set("Content-Type", "application/json")

	w := httptest.NewRecorder()
	s.ServeHTTP(w, r)

	assert.Equal(t, 0, len(p.msgs))
	assert.Equal(t, 403, w.Code)
	assert.Equal(t, "Forbidden\n", w.Body.String())
}

func TestServer_ServeHTTP_secret_correct(t *testing.T) {
	p := new(pub)

	s := Server{
		Log:       log.New(ioutil.Discard, "", log.LstdFlags),
		Topic:     "builds",
		Secret:    "wahoo",
		Publisher: p,
	}

	b := bytes.NewBufferString(`{ "foo": "bar" }`)

	r, err := http.NewRequest("POST", "/build?secret=wahoo", b)
	assert.Equal(t, nil, err)
	r.Header.Set("Content-Type", "application/json")

	w := httptest.NewRecorder()
	s.ServeHTTP(w, r)

	assert.Equal(t, 1, len(p.msgs))
	assert.Equal(t, 200, w.Code)
}

func TestServer_ServeHTTP_malformedJSON(t *testing.T) {
	p := new(pub)

	s := Server{
		Log:       log.New(ioutil.Discard, "", log.LstdFlags),
		Topic:     "builds",
		Publisher: p,
	}

	b := bytes.NewBufferString(`{ "`)

	r, err := http.NewRequest("POST", "/build", b)
	assert.Equal(t, nil, err)
	r.Header.Set("Content-Type", "application/json")

	w := httptest.NewRecorder()
	s.ServeHTTP(w, r)

	assert.Equal(t, 0, len(p.msgs))
	assert.Equal(t, 400, w.Code)
	assert.Equal(t, "Error parsing request body\n", w.Body.String())
}

func TestServer_ServeHTTP_health(t *testing.T) {
	p := new(pub)

	s := Server{
		Log:       log.New(ioutil.Discard, "", log.LstdFlags),
		Topic:     "builds",
		Publisher: p,
	}

	r, err := http.NewRequest("GET", "/internal/health", nil)
	assert.Equal(t, nil, err)

	w := httptest.NewRecorder()
	s.ServeHTTP(w, r)

	assert.Equal(t, 0, len(p.msgs))
	assert.Equal(t, 200, w.Code)
	assert.Equal(t, "OK\n", w.Body.String())
}

type responseWriterNoOp struct {
}

func (*responseWriterNoOp) Header() http.Header {
	return nil
}

func (*responseWriterNoOp) Write([]byte) (int, error) {
	return 0, nil
}

func (*responseWriterNoOp) WriteHeader(int) {
}

func BenchmarkServe(b *testing.B) {
	p := new(pub)

	s := Server{
		Log:       log.New(ioutil.Discard, "", log.LstdFlags),
		Topic:     "builds",
		Publisher: p,
	}

	for i := 0; i < b.N; i++ {
		r, err := http.NewRequest("POST", "/build", bytes.NewBufferString(`{ "foo": "bar" }`))
		if err != nil {
			b.Error(err)
		}
		r.Header.Set("Content-Type", "application/json")
		s.ServeHTTP(&responseWriterNoOp{}, r)
	}
}


cmd/http_to_nsq/main.go
package main

import (
	"log"
	"net/http"
	"os"

	"github.com/nsqio/go-nsq"
	"github.com/segmentio/http_to_nsq"
	"github.com/tj/docopt"
)

var Version = "0.2.0"

const Usage = `
  Usage:
    http_to_nsq --topic name
      [--nsqd-tcp-address addr]
      [--address addr]
      [--secret secret]
    http_to_nsq -h | --help
    http_to_nsq --version

  Options:
    --topic name             nsqd topic name
    --secret secret          secret string [default: ]
    --address addr           bind address [default: localhost:3000]
    --nsqd-tcp-address addr  nsqd tcp address [default: localhost:4150]
    -h, --help               output help information
    -v, --version            output version
`

func main() {
	args, err := docopt.Parse(Usage, nil, true, Version, false)
	if err != nil {
		log.Fatal("[error] %s", err)
	}

	topic := args["--topic"].(string)
	addr := args["--address"].(string)
	nsqd := args["--nsqd-tcp-address"].(string)
	secret := args["--secret"].(string)

	log.Printf("starting http_to_nsq %s", Version)
	log.Printf("--> binding to %s", addr)
	log.Printf("--> publishing to %s as topic %q", nsqd, topic)

	prod, err := nsq.NewProducer(nsqd, nsq.NewConfig())
	if err != nil {
		log.Fatal("[error] starting producer: %s", err)
	}

	server := &http_to_nsq.Server{
		Log:       log.New(os.Stderr, "", log.LstdFlags),
		Secret:    secret,
		Topic:     topic,
		Publisher: prod,
	}

	err = http.ListenAndServe(addr, server)
	if err != nil {
		log.Fatal("[error] binding: %s", err)
	}
}


server.go
package http_to_nsq

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
)

// Publisher.
type publisher interface {
	Publish(topic string, body []byte) error
}

// Message published to NSQD.
type Message struct {
	URL    string          `json:"url"`
	Method string          `json:"method"`
	Header http.Header     `json:"header"`
	Body   json.RawMessage `json:"body"`
}

// Server publishing requests as Messages.
type Server struct {
	Topic     string
	Secret    string
	Publisher publisher
	Log       *log.Logger
}

// ServeHTTP implements http.Handler.
func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	switch r.URL.Path {
	case "/internal/health":
		fmt.Fprintf(w, "OK\n")
	default:
		s.publish(w, r)
	}
}

// Publish requests to NSQD.
func (s *Server) publish(w http.ResponseWriter, r *http.Request) {
	if r.Method != "POST" && r.Method != "PUT" && r.Method != "DELETE" {
		s.Log.Printf("[error] invalid method: %s", r.Method)
		http.Error(w, http.StatusText(405), 405)
		return
	}

	secret := r.URL.Query().Get("secret")
	if s.Secret != "" && s.Secret != secret {
		s.Log.Printf("[error] invalid secret")
		http.Error(w, http.StatusText(403), 403)
		return
	}

	var body json.RawMessage
	err := json.NewDecoder(r.Body).Decode(&body)
	if err != nil {
		s.Log.Printf("[error] decoding body: %s", err)
		http.Error(w, "Error parsing request body", 400)
		return
	}

	msg := &Message{
		URL:    r.URL.String(),
		Method: r.Method,
		Header: r.Header,
		Body:   body,
	}

	b, err := json.Marshal(msg)
	if err != nil {
		s.Log.Printf("[error] marshalling message: %s", err)
		http.Error(w, http.StatusText(500), 500)
		return
	}

	s.Log.Printf("[info] publishing %s %s", msg.Method, msg.URL)
	err = s.Publisher.Publish(s.Topic, b)
	if err != nil {
		s.Log.Printf("[error] publishing body: %s", err)
		http.Error(w, http.StatusText(500), 500)
		return
	}

	fmt.Fprintf(w, ":)")
}
